cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java

```diff
@@ -240,6 +241,10 @@ protected void rebalanceForBroker(Broker broker,
         continue;
       }
 
+      if (onlyMoveImmigrantReplicas && !replica.isImmigrant()) {
```

Comment:
What if the (original) broker is dead and contains replicas? Then if `onlyMoveImmigrantReplicas` is true, this goal will be unable to move offline replicas away from it.

I feel that we can make the code cleaner and more concise if we can centralize the exclusion logic -- e.g. via a centralized `shouldExclude(Replica, OptimizationOptions)`.

Reply:
I think the logic in this patch so far is not considering moving offline replicas.
What this means is that if we have offline replicas in the cluster and specify to use hard goals when increasing RF,  the request will fail with `OptimizationException`.

To centralize the filtering logic, we can address that in PR #810

Reply:
If the expectation is to not run when there are offline replicas, can we fail early? -- i.e. we have the cluster model before starting the optimization.